#163388140 Authenticated users should be able to comment on articles

PaulinhoNdegwa opened the pull request:
#### What does this PR do?
It enables an authenticated user to comment on articles and reply to comments

#### Description of Task to be completed?
- Created comment model and serializer
- Creates comment CRUD views
- Written comment tests and refactored

#### How should this be manually tested?
1. Clone the repo

```
$ git clone https://github.com/andela/ah-the-phoenix.git
```
1. Change it into the current directory

```
$ cd ah-the-phoenix
```
Checkout to the Feature branch

```
$ git checkout ft-articles-comments-163388140
```
Run the server
   `python manage.py runserver`
1. Comment on an article
      Send a `POST` request to  `http://127.0.0.1:8000/api/v1/articles/<slug>/comment/`
      Sample comment:
```
{
"body": "This is a comment"
}
```
![image](https://user-images.githubusercontent.com/42174387/53070983-dbbefb80-34f1-11e9-9984-bcd71b658c9f.png)

2. GET all article comments
      Send a `GET` request to  `http://127.0.0.1:8000/api/v1/articles/<slug>/comment/`
![image](https://user-images.githubusercontent.com/42174387/53071075-35272a80-34f2-11e9-9436-90c3e166220f.png)

3. GET one comment
    Send a `GET` request to  `http://127.0.0.1:8000/api/v1/articles/<slug>/comment/<comment_id>`
![image](https://user-images.githubusercontent.com/42174387/53071069-2d678600-34f2-11e9-90d2-ba4f11910add.png)

4. UPDATE a comment
    Send a `PUT` request to  `http://127.0.0.1:8000/api/v1/articles/<slug>/comment/<comment_id>`
![image](https://user-images.githubusercontent.com/42174387/53071570-aa472f80-34f3-11e9-9a83-a4ec66e7f3d4.png)

5. DELETE a comment
    Send a `DELETE` request to  `http://127.0.0.1:8000/api/v1/articles/<slug>/comment/<comment_id>`
![image](https://user-images.githubusercontent.com/42174387/53071024-04df8c00-34f2-11e9-8aa6-c280e0bc3ba4.png)

6. REPLY to a comment
    Send a `POST` request to  `http://127.0.0.1:8000/api/v1/articles/<slug>/comment/<comment_id>`
    _* You cannot reply to a reply_
![image](https://user-images.githubusercontent.com/42174387/53071518-8683e980-34f3-11e9-815e-3dbd5497f2fc.png)


#### Any background context you want to provide?
None

#### What are the relevant pivotal tracker stories?

 #163388140

PaulinhoNdegwa committed:
feat(comments): Authenticated users should be able to comment on articles

- Created comment model and serializer
- Creates comment CRUD views
- Written comment tests and refactored

[Delivers #163388140]

Arusey commented on authors/apps/articles/migrations/0001_initial.py:
```diff
@@ -1,14 +1,17 @@
-# Generated by Django 2.1.5 on 2019-02-14 18:54
+# Generated by Django 2.1.5 on 2019-02-18 16:07
```
Nice implementation. good job

PaulinhoNdegwa replied:
Thanks.

sirghiny commented on authors/apps/articles/models.py:
```diff
@@ -41,3 +41,42 @@ def save(self, *args, **kwargs):
         super().save(*args, **kwargs)
+
+
+class Comments(models.Model):
```
Model names should be singular.

PaulinhoNdegwa replied:
Done

sirghiny commented on authors/apps/articles/views.py:
```diff
@@ -85,3 +86,145 @@ def destroy(self, request, pk=None):
+    def get_specific_comment(self, article_id, comment_id, request):
+        """This methos a single comment related to a specific article"""
+        try:
+            Article.objects.get(pk=article_id)
```
We'll need to abstract out the getting of common things e.g Articles in viewsets. You can leave it for now but change it in the code clean up.

PaulinhoNdegwa replied:
How do i abstract? Example in the articles scenario?

asirvex replied:
You can abstract the process of getting an article and returning an error if it is not found. In the latest changes there is a function called get_article that can do that

PaulinhoNdegwa replied:
Do you mean the `get_article` in the serializer?

PaulinhoNdegwa replied:
I will look into it

PaulinhoNdegwa replied:
I have made the changes on the final rebase commit..

asirvex commented on authors/apps/articles/views.py:
```diff
@@ -85,3 +86,145 @@ def destroy(self, request, pk=None):
+            comment = Comments.objects.filter(pk=comment_id,
+                                              article_id=article_id).first()
+        except Exception:
+            raise Http404("Data not found")
```
give a more specific error message

PaulinhoNdegwa replied:
Done

PaulinhoNdegwa committed:
feat(comments): Implement feedback

- Change model name to singular

{Delivers #163388140]

PaulinhoNdegwa committed:
feat(comments): Rebase from develop and refactor

- Rebase changes from develop
- Refactor changes and run tests

[Finishes #163388140]

Arusey reviewed (commented):
Looks good. Nice implementation

dev-jey commented:
Great job on the threading aspect of the comments@PaulinhoNdegwa

PaulinhoNdegwa commented:
> Great job on the threading aspect of the comments@PaulinhoNdegwa

Thanks @codact

PaulinhoNdegwa commented:
> Looks good. Nice implementation

Hello @Arusey, Kindly be more specific on what on I accomplished well. Thanks!
